virtual-programs/unix-commands.folk

```diff
@@ -1,33 +1,124 @@
-set ::unixjobs [dict create]
-set ::nextunixjobid 0
-
-proc ::readline {jobid channel} {
-	if {[gets $channel line] >= 0} {
-		dict with ::unixjobs $jobid {
-			lappend log $line
-			Retract $object is running Unix command $command with job id $jobid output /something/
-			Assert $object is running Unix command $command with job id $jobid output $log
-			Step
+# Spawns a Unix command to stream output lines back to the Wisher.
+#
+# Wish $p runs Unix command "echo" with arguments [list "Hello" "World"]
+# Wish $this runs Unix command "journalctl" with arguments [list "-f" "-u" "folk"]
+When /someone/ wishes /p/ runs Unix command /command/ with arguments /args/ {
+	set outputKeyName [list unix-output $p]
+	set errorKeyName [list unix-error $p]
+	set maxLines 500
+	set maxLinesEndIndex [expr {$maxLines - 1}]
+	set accumulatedLines [list]
+
+	# Bound how many lines we drain per tick to avoid starvation under heavy output
+	set maxLinesPerTick 10
+
+	# Build argv as a flat list and form pipeline tokens
+	set flatArgs [concat {*}$args]
+	set argv [list $command {*}$flatArgs]
+	set pipeline [linsert $argv 0 |]
+
+	try {
+		# Start the process with STDERR merged into STDOUT
+		lappend pipeline 2>@1
+		set fd [open $pipeline r]
+
+		fconfigure $fd -blocking 0 -buffering none
+
+		set pids [pid $fd]
+		set pid [lindex $pids end]
+	} on error e {
+		puts "Failed to open '$command': $e"
+
+		Hold! -key $errorKeyName \
+			Claim $p has Unix error output $e
+
+		return
+	}
+
+	On unmatch [list apply {{fd pid} {
+		catch {close $fd}
+		catch {kill SIGTERM $pid}
+		after 500
+		catch {kill SIGKILL $pid}
+	} } $fd $pid]
+
+	while true {
+		set newLines [list]
+		set drained 0
+
+		while {$drained < $maxLinesPerTick} {
+			set num [gets $fd line]
+			if {$num < 0} { break }
+
+			lappend newLines $line
+			incr drained
+		}
+
+		if {[llength $newLines] > 0} {
+			set accumulatedLines [concat $accumulatedLines $newLines]
+			set length [llength $accumulatedLines]
+
+			if {$length > $maxLines} {
+				set accumulatedLines [lrange $accumulatedLines end-$maxLinesEndIndex end]
+			}
+
+			Hold! -key $outputKeyName \
+				Claim $p has Unix output lines $accumulatedLines
```

Comment:
Oh, I also wonder if we should scope these output/error claims in case you want to run multiple commands on the same $p. We might need an ugly job-id-like thing for that?

Reply:
I'll work on that, good idea! Can I just use the `$pid`?